### PR TITLE
Add theme choose in footer

### PR DESCRIPTION
**Pull request by a1012112796**
As title.
which will allow user change theme in anywhere by a easy way.
Use cookies to implement this.

### CODE REVIEW

**Comment by lafriks** [comment]
shouldn't this also be post request like language?

**Comment by techknowlogick** [comment]
> shouldn't this also be post request like language?

yes, as otherwise GET requests can change user info, which can be used as a CSRF attack (ex. ``)

**Comment by a1012112796** [comment]
> > shouldn't this also be post request like language?
> 
> yes, as otherwise GET requests can change user info, which can be used as a CSRF attack (ex. ``)

Hmm. I'm sorry, but I can't fully understand you. This link will only set cookie. Why can it change user info? Thanks

**Comment by silverwind** [comment]
I think instead of creating a second theme variable, we should consolidate to a single `.Theme` one. Signed in users should store their theme server-side as usual (current `.SignedUser.Theme` variable) and unauthenticated users could use the cookie approach. Make sure to update all uses of `.SignedUser.Theme` to `.Theme`.

**Comment by silverwind** [comment]
Instead of putting it into the already crammed footer, I'd suggest putting it into the user button menu as a second-level dropdown. See https://fomantic-ui.com/modules/dropdown.html for an example:

image

**Comment by 6543** [comment]
I'm against using unessesary cockies anyway ...

**Comment by silverwind** [comment]
Agree, cookies should be avoided but it's kind of hard when the server has nothing in the database about a unauthenticed user's theme preference and you don't want to use a JS-based solution. If we go ahead with the cookie, we should ensure it's deleted on login.

**Comment by a1012112796** [comment]
> I'm against using unessesary cockies anyway ...

Hmm, that's because I'd like to allow users to choose the current theme without login.

**Comment by a1012112796** [comment]
> Agree, cookies should be avoided but it's kind of hard when the server has nothing in the database about a unauthenticed user's theme preference and you don't want to use a JS-based solution. If we go ahead with the cookie, we should ensure it's deleted on login.

In my opinion, the theme in user settings is default setting for this user, which should be used when user does not have current setting in this terminal's cookie. But if have, I think the current setting should have higher priority than default setting. Because it's a new but temporary setting. Thanks.

**Comment by a1012112796** [comment]
> Instead of putting it into the already crammed footer, I'd suggest putting it into the user button menu as a second-level dropdown. See https://fomantic-ui.com/modules/dropdown.html for an example:
> 
> image

It can't be used when users do not login....

**Comment by a1012112796** [comment]
> I think instead of creating a second theme variable, we should consolidate to a single `.Theme` one. Signed in users should store their theme server-side as usual (current `.SignedUser.Theme` variable) so it works across devices and unauthenticated users could use the cookie approach. Make sure to update all uses of `.SignedUser.Theme` to `.Theme`.

As mentioned earlier, I think there is a difference between the user's current setting and the user's default setting. The current setting is only the theme required by the user's current time and device. The user's default setting should not be changed because of this. The user default Settings still need to be changed through the user Settings page. Thanks

File: routers/routes/routes.go [hunk 1]
```diff
@@ -277,6 +278,60 @@ func RegisterRoutes(m *macaron.Macaron) {
...
+				if strings.EqualFold(strings.ToLower(v), strings.ToLower(chooseTheme)) {
+					exists = true
+					break
+				}
+			}
+
+			if exists {
+				themeNow = chooseTheme
+			} else {
+				ctx.Flash.Error(ctx.Tr("settings.theme_update_error"))
```
Comment (bagasme) [inline routers/routes/routes.go]: What about the error message itself?

**Comment by silverwind** [comment]
> It can't be used when users do not login....

Maybe move the dropdown besides the user/signin button?

As for an icon, it seems the octicons do not have a suitable one, but we can embed this one in the HTML (should be replaced with a better mechanism to load custom SVGs later):

https://raw.githubusercontent.com/google/material-design-icons/master/image/svg/production/ic_palette_48px.svg

**Comment by CirnoT** [comment]
I don't agree with putting it in header, it's not used often enough to warrant wasting space there.

File: routers/routes/routes.go (hunk 1 above)
Comment (a1012112796) [reply routers/routes/routes.go]: https://github.com/go-gitea/gitea/blob/209b17c4e25fe72d2fdf46f412fc388bc274a516/options/locale/locale_en-US.ini#L473

**Comment by silverwind** [comment]
I just find it a bit odd to have controls down there, but we already have the language dropdown so it probably is fine.

**Comment by silverwind** [comment]
> As mentioned earlier, I think there is a difference between the user's current setting and the user's default setting. The current setting is only the theme required by the user's current time and device. The user's default setting should not be changed because of this. The user default Settings still need to be changed through the user Settings page.

I guess some rare user might prefer a dark theme on one device and a light one on another, that's fine. I'm more concerned with the multiple variables thought, they should be combined into one `.Theme` imho.

File: routers/user/setting/account.go [hunk 2]
```diff
@@ -262,6 +262,7 @@ func UpdateUIThemePost(ctx *context.Context, form auth.UpdateThemeForm) {
 		return
 	}
 
+	ctx.SetCookie("themeNow", ctx.User.Theme, nil, setting.AppSubURL, setting.SessionConfig.Domain, setting.SessionConfig.Secure, true)
```
Comment (silverwind) [inline routers/user/setting/account.go]: Maybe name it `current_theme` which would be more syntactical correct and follow the casing convention of other cookies like `i_like_gitea`.

**Comment by silverwind** [comment]
1. I get a persistent error on the frontpage:

image

2. There is a border issue:

image

 Issue is that dropdowns are block level white anchors are inline level. I'd just remove these borders by removing these rules:

````
footer .container .links > * {
    border-left-color: #888;
}

footer .container .links > * {
    border-left: 1px solid #d6d6d6;
}
````

3. Maybe split out the cookie stuff into a separate PR so this one only adds the dropdown? Would be easiert to land because that cookie part is kind of controversial.

**Comment by CirnoT** [comment]
I think the point of this PR is to have ability for user to use different theme on different computers.

**Comment by a1012112796** [comment]
> 1. I get a persistent error on the frontpage (dropdown shows `arc-green`):
> 
> image
> 
> 1. There is a border issue:
> 
> image
> 
> Issue is that dropdowns are block level white anchors are inline level. I'd just remove these borders by removing these rules:
> 
> ```css
> footer .container .links > * {
>     border-left-color: #888;
> }
> 
> footer .container .links > * {
>     border-left: 1px solid #d6d6d6;
> }
> ```
> 
> 1. Maybe split out the cookie stuff into a separate PR so this one only adds the dropdown for logged in users? Would be easier to approve because that cookie part is kind of controversial.

Thanks for your help. Using cookies to save the theme choose is the main change for this pr, if no one thinks it's meaningful, this pr can be closed, Thanks.

**Comment by a1012112796** [comment]
Close this pr because after carefully thinking, I think it's not a meaningful change. Thanks

**Comment by a1012112796** [comment]
@lafriks  Please remove the Milestone set, Thanks, Thanks.

**Pull request closed without merging**